internal/memory/types.go

```diff
@@ -0,0 +1,34 @@
+package memory
+
+import "github.com/pkg/errors"
+
+type classType uint32
+
+const (
+	MegaByte = uint64(1024 * 1024)
+	GigaByte = 1024 * MegaByte
+)
+
+var (
+	ErrNotEnoughSpace = errors.New("not enough space")
+	ErrNotAllocated   = errors.New("no memory allocated at the given offset")
+)
+
+// mappedRegion represents a memory block with an offset
+type MappedRegion interface {
+	Offset() uint64
+	Size() uint64
+}
+
+// allocator takes care of memory management
+//
+// The main use case for allocator is tracking memory regions allocated on a Virtual PMem
+// device, when mapped devices feature is being used.
+type Allocator interface {
+	Allocate(classType) (MappedRegion, error)
+	Expand(classType) error
+	Release(classType, uint64) error
+
+	Busy() bool
+	FindFreeOffset(classType) (uint64, error)
```

Comment:
I am still trying to read through the entire PR (so I might be missing something) but I think `Allocator` interface shouldn't really have `Busy` and `FindFreeOffset` methods. Checking if a region is busy is not the job of an allocator (even though it needs to do that when allocating a new region, but that's an implementation detail), similarly finding free offset is an implementation detail. What do you think?

Reply:
yeah, now that I have the `merge` logic (which didn't exist in early iterations), maybe I can get away with not exposing `FindFreeOffset` and rather do `Allocate/defer Release` 🤔 and you're right that `Busy` is probably something that should be part of `vPMemMultiInfo` or something.

Reply:
Yep, similarly we might be able to remove `Expand` also I guess since that's an implementation detail.